### ISSUE TITLE
Promote a common approach to storing Github and Gitlab credentials

### ISSUE THREAD

**Issue by tarsius** [magit/magit#2973]
I believe there are currently two Elisp libraries for accessing the Github api; [`gh`](https://github.com/sigma/gh.el) and my [`ghub`](https://github.com/tarsius/ghub).  And I am sure some end-user packages come with their own tools for accessing the api.  This is all perfectly fine, but for users who use multiple packages by different authors it is inconvenient to have to setup the credentials multiple times.  In `ghub` I made an effort to leverage existing Emacs features instead of reinventing the wheel.

These are the setup instructions:

```
;; Initial configuration:
;;
;;   $ git config github.user <username>
;;   $ emacs ~/.authinfo.gpg
;;   # -*- epa-file-encrypt-to: ("A.U.Thor@example.com") -*-
;;   machine api.github.com login <login> password <token>
```

Note very chatty, but quite self-explanatory I hope.  And all the code related to this in `ghub.el` is this:

```lisp
(require 'auth-source)

(defun ghub--get-access-token ()
  (let ((secret
         (plist-get (car (auth-source-search
                          :max 1
                          :user (substring (shell-command-to-string
                                            "git config github.user")
                                           0 -1)
                          :host ghub--domain))
                    :secret)))
    (if (functionp secret)
        (funcall secret)
      secret)))
```

(Note how I did not even use Magit when calling `git' ;-)

I have seen other packages doing essentially the same in a much more complicated fashion, for both the user as well as the author.  You don't need to depend on `ghub` to do use its approach, it might not be to your taste (and is subject to change (likely only better error handling, but the use of `libcurl` is also on the table)).  Please feel free to copy-paste this snippet into your package instead.

----

Of course the same applies to Gitlab (my `glab.el` is almost identical to `ghub.el` except `s/ghub/glab/`), and Bitbucket.

----

Please also note that I have not yet invited any of the involved parties to this conversation because I am currently trying to outline my future work on Magit and opening many issues in the process.  Only once I am done with that, will I actually try to get this conversation going.

**Comment by tarsius** [comment]
We could probably get away with the function case. I don't remember to what extend I investigated whether `auth-source-search` will actually ever return a function given our configuration. I'll do that at some point.

**Comment by tarsius** [comment]
Thinking about it again, the above "setup instructions" probably aren't enough. We probably need something that is more convenient to end-users. I have heard that `gh.el` now does but haven't looked at that yet.

On the other hand at this point `gh.el` is mainly intended for my own use - I use it to maintain the Emacsmirror, so for that package it is okay to require some manual setup (because essentially one person had to do it once ;-)

The argument I am trying to make doesn't so much concern how a package should semi-automatically get the required information or guide the user in providing it by editing a file. My point is that it should be stored the same way by all packages and that the right way to do it is `auth`.

**Comment by blaenk** [comment]
The way I like to do it—and I'm not necessarily advocating that this be done for all packages like you're saying—is to have this at the end of my `~/.gitconfig`:

```
[include]
	path = ~/.gitconfig.local
```

Then my `~/.gitconfig.local` file contains any sensitive or personal information that I don't want to include in my dotfiles repository. The advantage of this is that then any application, be it a package in emacs or the `hub` command for example, can find and use this token, because it can be picked up through `git config`.

This is what `gh.el` currently does, where it probes `git config` to see if the oauth token is already present there, and if not then it continues on with the oauth token retrieval flow.

My point is that at the very least I think people who do something like what I do would appreciate it if `git config` were consulted for the presence of the token before consulting `auth-source`, or vice versa, but that `git config` is checked at all would be nice, so that people who already have a way of handling this information don't have to repeat it in yet another way.

**Comment by blaenk** [comment]
With respect to being very convenient to end-users, I think it would indeed be desirable to make this as hands-off as possible. `gh.el` does do something like this except they simply write (via `git config`) the information to `~/.gitconfig`. If that isn't wanted, another idea perhaps is to save it in some `.github.credentials`-type file in the `~/.emacs.d/` directory.

If it's possible to encrypt that within emacs in a hands-off manner that doesn't require the user to create and supply a key then go for it, but I don't think it's necessary to be honest. The oauth token is already stored in `git config` by convention, the token should have limited permission scopes, and it's something that can simply be revoked in the worst case, it's not like storing the username and password.

**Comment by vermiculus** [comment]
I think storing the authentication information in `.authinfo` makes the most sense moving forward. I'd be surprised if any major git collaboration tool didn't provide a token-based password in some respect, but it's not out of the question. Personally, I wouldn't want my plaintext password sitting around *anywhere* -- the ability to encrypt it and have Emacs just 'know what to do' is extremely attractive to me.

**Comment by blaenk** [comment]
Storing a password, plain-text or not, should be out of the question. Thankfully no one has suggested that.

**Comment by vermiculus** [comment]
That's exactly what `gh.el` does, though. (Well, technically it's an auth token, but I view it with a similar caution.)

**Comment by blaenk** [comment]
One can limit tokens to specific scopes of permissions, such as only for gists, etc. Even GitHub's own [hub](https://github.com/github/hub) utility stores the auth token in a plain-text `~/.config/hub` file. I'm not saying we should therefore blindly follow what they do (though other utilities also do and expect this), but I don't think it's fair to characterize storing a password as being the same as storing a scope-limited auth token.

In the event that a token is compromised it would only grant access to the permissions associated with the token, and it can simply be revoked while everything else still works. A compromised password on the other hand gives access to everything on the account, including all existing auth tokens.

I'm all for better-safe-than-sorry security and encryption, I only say this given the interest in also balancing for ease-of-use for the end-user. If the authinfo workflow can't be streamlined to a degree you/we're comfortable with, it's probably not the end of the world to simply store the auth token like other utilities do.

**Comment by vermiculus** [comment]
> If the authinfo workflow can't be streamlined to a degree you/we're comfortable with, it's probably not the end of the world to simply store the auth token like other utilities do.

Absolutely agree.

**Comment by tarsius** [comment]
Status update.

* `ghub.el` is mostly ready, but see https://github.com/magit/ghub/issues/34.
* A new `glab.el` implemented on top of `ghub.el` is also in the works. See https://github.com/magit/ghub/pull/49.

I have to finish those things up and create releases and announce those. That concludes the "promote" part.

I have decided to not do the same for Bitbucket because it is neither FOSS nor used by many Emacs users.

**Comment by tarsius** [comment]
Tadaaa https://emacsair.me/2018/03/20/ghub-2.0.

**Comment by DrWaleedAYousef** [comment]
I  spent hours trying to let magit get the user name and password from `authinfo.gpg` I failed. However, I managed for magithub. My .emacs file is:
```
(use-package magit)
(use-package ghub);; for token storing
(use-package magithub
  :after magit
  :ensure t
  :config (magithub-feature-autoinject t))

(setq auth-sources '((:source "~/.authinfo.gpg")))
(setq magit-process-find-password-functions '(magit-process-password-auth-source))
```

my `authinfo.gpg` file is:
```
machine api.github.com login DrWaleedAYousef^magithub password TOKEN

machine api.github.com login DrWaleedAYousef^ghub password TOKEN
```

**Comment by tarsius** [comment]
> `(setq auth-sources '((:source "~/.authinfo.gpg")))`

This is wrong, it should be `(setq auth-sources '("~/.authinfo.gpg"))`.

> my `authinfo.gpg` file is:

I suspect this is a typo in your post only but this is wrong too -- the name of the file is `.authinfo.gpg`.

Magit and Magithub both use `ghub`, so if it works for one of them it should work for the other. In the `.authinfo.gpg` you posted there is an entry for `magithub` and one for `ghub`. To give `magit` access, there also has to be an entry for `magit`. The easiest way to accomplish this probably is to copy the `magithub` line and then remove the `hub` from the copy.

**Comment by DrWaleedAYousef** [comment]
I really appreciate your fast and detailed response. I was wrong in `setq', I modified it. the `authinfo.gpg` is just a typo in the email. Still it asks me for the username and password; however, I noticed that (when I do `P u` to push) emacs reads in the minibuffer the `.authinfo.gpg`. Nevertheless, it asked me for the password. For clarity, by `.emacs` and `.authinfo.gpg` are:
```
(use-package magit)
(use-package ghub);; for token storing
(setq auth-sources '("~/.authinfo.gpg"))
(setq magit-process-find-password-functions '(magit-process-password-auth-source))
;; I commented out the magithub package for easier tracing
```

```
machine api.github.com login DrWaleedAYousef^magit password TOKEN1

machine api.github.com login DrWaleedAYousef^ghub password TOKEN2
```
I tried TOKEN1 and TOKEN2 to be equal and another time to be different.

Thanks so much.

**Comment by tarsius** [comment]
> Nevertheless, it asked me for the password. 

To read the contents of `.authinfo.gpg` Emacs has to ask you for your gpg passphrase once per session.

> I tried TOKEN1 and TOKEN2 to be equal and another time to be different.

Did either one of them work? (It shouldn't matter which approach you use.)

By the way try to avoid empty lines or even comments in `.authinfo.gpg`. There are some bugs in `auth-source` that are triggered by such lines. `ghub` patches those bugs and they are fixed in Emacs 26.1.50, but avoid such lines anyway, just in case.

**Comment by DrWaleedAYousef** [comment]
Thanks so much again

It is not the gpg passphrase of `.authinfo.gpg' (I am familiar with that since I am using gnus). It is indeed the username and password of Git; the message is:
```
Username for 'https://github.com': 
Password for 'https://DrWaleedAYousef@github.com': 
```
I tried to enter them once; then tried again, e.g., `F u` it asked again! I tried removing all empty lines. I tried as well to put the two lines of `magit` and `ghub` at the beginning; nothing working!!

You know, I have been trying for this now since yesterday, reading dozens of webpages. 

Thanks a million for your patience; looking forward.

**Comment by tarsius** [comment]
On https://github.com/vermiculus/magithub/issues/360 you said:

> After magit-status I do magit-gh-pulls-mode,

`magit-gh-pulls` is a *third* package beside `magit` and `magithub` and it uses a different configuration mechanism. The above prompts are likely from that package.

**Comment by DrWaleedAYousef** [comment]
That was my first attempt. When I did not receive more responses from him,
I tried to remove magit, magithub, pops,...and do everything cleaner.

Then I just restarted emacs to install magit and magithub using the
use-package statements. I no longer use magit-gh-*. The only emacs
configurations I have are those I shared with you.

The weired 2 things are: 1. magithub is working fine (except when I choose
to see a comment/issue on the browser; it denies credentials. Probably this
is the normal behavior) 2. magit indeed opens .authinfo.gpg as I see in the
mini buffer; but then it asks for credentials!

Thanks so much.

**Comment by DrWaleedAYousef** [comment]
Are you still interested in troubleshooting this? I know you may be very
busy. Thanks in advance.

**Comment by tarsius** [comment]
I have just added a new command `ghub-clear-caches`, whose purpose is to get rid of cached but incorrect information. But it doesn't accomplish anything that restarting Emacs wouldn't do too, so it probably won't help here.

**Comment by tarsius** [comment]
> I tried to enter them once; then tried again, e.g., `F u` it asked again!

Aha! That doesn't make use of `ghub` and `auth-source` at all. Pulling is called by calling `git pull ...`, so all the configuration that is necessary to get information using the Github API is irrelevant here.

You have to separately configure `git` to teach it the secrets it needs to talk to the remote repository. You have probably already done so (i.e. you are able to push and pull on the command line just fine), and the reason that it doesn't work with Magit is that you have done so in a way that is visible in a shell, but not inside Emacs.

As a quick check run `emacs &` in a shell instead of doing whatever you usually do to start Emacs. If that works, then you have to figure out how to always make your configuration known to Emacs.

Some things that might help:

* https://github.com/magit/magit/wiki/Pushing-with-Magit-from-Windows
* https://github.com/purcell/exec-path-from-shell
* https://github.com/tarsius/keychain-environment

> Are you still interested in troubleshooting this?

Well, I am not so happy that you hijacked an old issue, and we generally don't want to use the issue tracker for support requests at all. But since we have gone this far...

Anyway, I don't think this is a defect. Maybe the documentation could be clearer and all, but I think the issue is a misconfiguration (or missing configuration) on your part.

**Comment by DrWaleedAYousef** [comment]
Thanks so much;

1. It was shocking for me to know that I have to configure `git` to read the `.authinfo.gpg`. I thought it is the responsibility of `magit` to read it and communicate the information to `git`; this is how I got the feeling from the manual page. Otherwise, whey then `magit` reads `.authinfo.gpg`?

2. Since I received your reply above, I spent hours reading how to configure `git` to read the `.authinfo.gpg`; I did not find it! I know it is out of your scope, but if it takes from you just few seconds to send me the configuration it would be great.

Thanks so much again, and sorry that I "hijacked" the issue. I thought it is the natural place since it is so related to me problem.

**Comment by tarsius** [comment]
1. Magit reads `.authinfo.gpg` to retrieve the secret needed to communicate with Github's proprietary API. This allows Magit to retrieve a list of open issues and pull-requests for example. Git is not the same thing as Github. Git does not have a built-in issue tracker. Magit uses the `ghub` package (which is what ultimately reads `.authinfo.gpg`) to communicate with the parts of Github that are not plain old Git.

2. To communicate with a remote `git` repository (whether hosted on Github or elsewhere) it uses `git`. `git` does not use `.authinfo.gpg`. To learn how to configure `git` so that it can talk to Github. https://help.github.com/articles/connecting-to-github-with-ssh might help, but there many other good resources out there.

3. I recommend you get this going on the command-line first before making it work with Magit. Probably completely leave "magit" out of your search queries and till you understand how `git` by itself has to be configured.

**Comment by DrWaleedAYousef** [comment]
Thats so helpful and informative; I will try do the homework. One last
thing before closing this (and probably it was part of the confusion) is
this. What you mentioned in 1. below is exactly what magithub does. What is
the difference then?

**Comment by tarsius** [comment]
Magithub provides many more features related to issues and pull-requests than Magit does. Essentially Magit currently only provides some commands that allows you to create a new local branch that contains the commits that a pull-request on Github has. These commands are bound to `b y` and `b Y`. Magithub on the other hand also shows a list of issues and pull-requests in Magit's status buffer and you can press `RET` on such an issue to show the related conversation.

**Comment by DrWaleedAYousef** [comment]
I really appreciate your patience, support, and time. Thank you for your
great package. At the end (as feature request), it would be great:
1. If magit provides scripts/config to auto configure the .authinfo
communication with git that confused me and still I have to search to
configure it.

2. If either magit and magithub merge together or magit provides the same
functionalities.

Thanks so much once again and please feel free to close this (I am sending
this from my cell with no access to Github).

--Waleed.

**Comment by vermiculus** [comment]
> If either magit and magithub merge together or magit provides the same
> functionalities.

That's not out of the question (it's something of an end-goal for me as well), but I think Jonas and I are on the same page that magithub's functionality is not yet mature enough to incorporate directly into magit. I view magithub as more of an experiment at that may be possible; magit has other considerations (like not providing functionality for one 'forge' over another) that require more careful and deliberate thought.
